src/sinks/util/sink.rs

```diff
@@ -520,16 +522,20 @@ where
     fn poll_complete(&mut self, cx: &mut Context<'_>) -> Poll<()> {
         while !self.in_flight.is_empty() {
             match ready!(Pin::new(&mut self.in_flight).poll_next(cx)) {
-                Some(Ok((seqno, batch_size))) => {
-                    self.pending_acks.insert(seqno, batch_size);
-
-                    let mut num_to_ack = 0;
-                    while let Some(ack_size) = self.pending_acks.remove(&self.seq_tail) {
-                        num_to_ack += ack_size;
+                Some(Ok((seqno, num_items, num_bytes))) => {
+                    self.pending_acks.insert(seqno, (num_items, num_bytes));
+
+                    let mut total_items = 0;
+                    let mut total_bytes = 0;
+                    while let Some((num_items, num_bytes)) =
+                        self.pending_acks.remove(&self.seq_tail)
+                    {
+                        total_items += num_items;
+                        total_bytes += num_bytes;
                         self.seq_tail += 1
                     }
-                    trace!(message = "Acking events.", acking_num = num_to_ack);
-                    self.acker.ack(num_to_ack);
+                    trace!(message = "Acking events.", acking_num = total_items);
+                    (self.on_success)(total_items, total_bytes);
                 }
```

Comment:
I think this logic points to the need for separate `acker` and `on_success` parameters. When an in flight request is complete, we should be emitting the event immediately (via `on_success`). The `pending_acks` logic is there to ensure the acknowledgements are issued strictly in sequence. I don't think those two parts should be joined.